src/emitters/astroobject.cpp

```diff
@@ -0,0 +1,256 @@
+#include <mitsuba/core/bsphere.h>
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/warp.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/scene.h>
+#include <mitsuba/render/texture.h>
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**!
+
+.. _emitter-astroobject:
+
+Distant astroobject emitter (:monosp:`astroobject`)
+---------------------------------------------------
+
+.. pluginparameters::
+
+ * - irradiance
+   - |spectrum|
+   - Spectral irradiance, which corresponds to the amount of spectral power
+     per unit area received by a hypothetical surface normal to the specified
+     direction.
+   - |exposed|, |differentiable|
+
+ * - to_world
+   - |transform|
+   - Emitter-to-world transformation matrix.
+   - |exposed|
+
+ * - direction
+   - |vector|
+   - Alternative (and exclusive) to `to_world`. Direction towards which the
+     emitter is radiating in world coordinates.
+
+ * - angular_diameter
+   - |float|
+   - Angular diameter of the astroobject in degrees.
+
+This emitter plugin implements a distant astroobject source which radiates a
+specified power per unit area within a solid angle. This solid angle is the
+proportion of the FOV taken by the emitter. This is parameterized by an
+angular_diameter. By default, the emitter radiates in the direction of the
+positive Z axis, i.e. :math:`(0, 0, 1)`, with an angular diameter of about the
+equivalent average sun angular diameter from earth.
+
+.. tabs::
+    .. code-tab:: xml
+        :name: astroobject-light
+
+        <emitter type="astroobject">
+            <vector name="direction" value="1.0, 0.0, 0.0"/>
+            <rgb name="irradiance" value="1.0"/>
+        </emitter>
+
+    .. code-tab:: python
+
+        'type': 'astroobject',
+        'direction': [1.0, 0.0, 0.0],
+        'irradiance': {
+            'type': 'rgb',
+            'value': 1.0,
+        }
+
+*/
+
+#define SUN_ANGULAR_DIAMETER 0.5358f
+
+MI_VARIANT class AstroObjectEmitter final : public Emitter<Float, Spectrum> {
+public:
+    MI_IMPORT_BASE(Emitter, m_flags, m_to_world, m_needs_sample_3)
+    MI_IMPORT_TYPES(Scene, Texture)
+
+    AstroObjectEmitter(const Properties &props) : Base(props) {
+        /* Until `set_scene` is called, we have no information
+           about the scene and default to the unit bounding sphere. */
+        m_bsphere = ScalarBoundingSphere3f(ScalarPoint3f(0.f), 1.f);
+
+        if (props.has_property("direction")) {
+            if (props.has_property("to_world"))
+                Throw("Only one of the parameters 'direction' and 'to_world' "
+                      "can be specified at the same time!'");
+
+            ScalarVector3f direction(dr::normalize(props.get<ScalarVector3f>("direction")));
+            m_direction = direction;
+            auto [up, unused] = coordinate_system(direction);
```

Comment:
Too bad C++ still doesn't allow to use std::ignore as part of structured bindings. A cleaner way to do this would be:
```suggestion
            ScalarVector3f direction(dr::normalize(props.get<ScalarVector3f>("direction"))), up;
            std::tie(up, std::ignore) = coordinate_system(direction);
```